enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+
+The proposed solution is based on pre-loading and pinning all required images
+in all the nodes of the cluster before starting the upgrade, and ensuring that
+no component requires access to a registry server during the upgrade. For this
+to work the following changes are required:
+
+1. No OpenShift component used during the upgrade should use the `Always` pull
+policy, as that forces the kubelet and CRI-O to try to contact the registry
+server even if the image is already available.
+
+1. No OpenShift component should garbage collect the images required for the
+upgrade. This is typically started by the kubelet instructing CRI-O to remove
+images.
+
+1. No OpenShift component should explicitly try to contact the registry server
+without a fallback alternative.
+
+1. The machine config operator needs to support image pinning, pre-loading and
+reloading of the CRI-O service.
+
+1. The cluster version operator needs to orchestrate the upgrade process.
+
+1. The engineer that prepares the upgrade bundle needs a `oc adm upgrade create
+bundle` tool to create it.
+
+1. The technician that performs the upgrade needs documentation explaining how
+to use the upgrade bundle.
+
+#### Don't use the `Always` pull policy during the upgrade
+
+Some OCP core components currently use the `Always` image pull policy during the
+upgrade. As a result, the kubelet and CRI-O will try to contact the registry
+server, even if the image is already available in the local storage of the
+cluster. This blocks the upgrade.
+
+The catalog operator uses the `Always` pull policy to pull catalog images. It
+does so in order refresh catalog images that are specified with a tag. But it
+also does it when it pulls catalog images that are specified with a digest. That
+should be changed to use the `IfNotPresent` pull policy for catalog images that
+are specified by digest.
+
+Most OCP core components have been changed in the past to avoid this use of the
+`Always` pull policy. Recently the OVN pre-puller has also been changed (see
+this [bug](https://issues.redhat.com/browse/OCPBUGS-13219) for details). To
+prevent bugs like this happening in the future and make the solution less
+fragile we should have a test that gates the OpenShift release and that
+verifies that the upgrade can be performed without a registry server. One way
+to ensure this is to run in CI an admission hook that rejects/warns about any spec
+that uses the `Always` pull policy.
+
+It would also be useful to have another test that scans for use of this `Always`
+pull policy in the source code.
+
+We can control the image pull policy for the OCP payload, but not for
+customer-specific images. The OCP upgrade may succeed, but the overall upgrade
+process will still be seen as failed from the customer point of view. To
+prevent that the upgrade mechanism will verify that there are no existing pods
+with the `Always` pull policy, and will abort the upgrade if there are any.
```

Comment:
The OCP upgrade itself wouldn't be blocked because of failures to pull images for customer workloads. But to the customer the overall upgrade would have failed because her workloads are no longer working.

In the OpenShift appliance use case (which is where we started) both the cluster and the customer workloads are installed using a temporary registry server that is shutdown after the installation is complete. That assumption doesn't work in general. I will change this to a warning.

Reply:
> In the OpenShift appliance use case (which is where we started) both the cluster and the customer workloads are installed using a temporary registry server that is shutdown after the installation is complete. That assumption doesn't work in general. I will change this to a warning.

Can you call this detail out somewhere in the enhancement, it seems important and helps someone understand when they don't have all the context (me!)

Reply:
Added that reference to this paragraph.